gcloud-sdk/src/token_source/ext_creds_source/mod.rs

```diff
@@ -139,3 +175,186 @@ fn subject_token_from_json(
+        regional_cred_verification_url: &str,
+        credentials: Credentials,
+        region: Region,
+        sign_at: SystemTime,
+        audience: &str,
+    ) -> crate::error::Result<SecretValue> {
+        let identity = credentials.into();
+        let signature_time = sign_at;
+
+        let mut signing_settings = SigningSettings::default();
+        signing_settings.signature_location = SignatureLocation::Headers;
+        let v4_signing_params = V4SigningParams::builder()
+            .name("sts")
+            .identity(&identity)
+            .region(region.as_ref())
+            .time(signature_time)
+            .settings(signing_settings)
+            .build()
+            .map_err(|e| Error::from(ErrorKind::ExternalCredsSourceError(e.to_string())))?;
+        let params = SigningParams::V4(v4_signing_params);
+
+        let regional_cred_verification_url =
+            regional_cred_verification_url.replace("{region}", region.as_ref());
+        let subject_token_url = regional_cred_verification_url;
+        let url = url::Url::parse(&subject_token_url).unwrap();
+        let method = Method::POST;
+        let mut headers = vec![("x-goog-cloud-target-resource", audience)];
+        if let Some(host) = url.host_str() {
+            headers.push(("Host", host))
+        }
+        let mut req = Request::builder().uri(url.to_string()).method(&method);
+        for header in &headers {
+            req = req.header(header.0, header.1);
+        }
+        let mut request = req.body(()).unwrap();
+
+        let signable_request = aws_sigv4::http_request::SignableRequest::new(
+            "POST",
+            &subject_token_url,
+            headers.into_iter(),
+            SignableBody::empty(),
+        )
+        .map_err(|e| Error::from(ErrorKind::ExternalCredsSourceError(e.to_string())))?;
+        let (instruction, _) = aws_sigv4::http_request::sign(signable_request, &params)
+            .map_err(|e| Error::from(ErrorKind::ExternalCredsSourceError(e.to_string())))?
+            .into_parts();
+        instruction.apply_to_request_http1x(&mut request);
+        let payload = AWSRequest {
+            url: subject_token_url.to_string(),
+            method: method.to_string(),
+            headers: request
+                .headers()
+                .into_iter()
+                .flat_map(|(k, v)| {
+                    v.to_str()
+                        .ok()
+                        .map(|v| AWSRequestHeader::new(k.to_string(), v.to_string()))
+                })
+                .collect(),
+        };
+        let payload =
+            serde_json::to_string(&payload).map_err(|e| Error::from(ErrorKind::TokenJson(e)))?;
+        let sts_token = utf8_percent_encode(&payload, NON_ALPHANUMERIC).to_string();
+
+        Ok(sts_token.into())
+    }
+
+    pub async fn get_aws_props() -> crate::error::Result<(Credentials, Region)> {
+        let imds_region_provider = ImdsRegionProvider::builder().build();
```

Comment:
`get_aws_props` uses providers from aws sdk that encapsulate credentials and metadata lookup-mechanism. It does not correspond one-on-one to AIP 4117, but If we use the provider from aws_config first and then fallback to imds provider, it covers most parts of preprocess of subject token generation.

The following bullet list is sited from https://google.aip.dev/auth/4117


- Validate the host for the url, regional_url and imdsv2_session_token_url fields if they are provided. The host should either be 169.254.169.254 or fd00:ec2::254.
- If imdsv2_session_token_url is available, then fetch session token from imdsv2_session_token_url.
- Check the environment variables in the following order (AWS_REGION and then the AWS_DEFAULT_REGION) to determine the AWS region. If found, skip using the AWS metadata server to determine this value.
- If the region environment variables are not provided, use the region_url to determine the current AWS region. The API returns the zone name, e.g. us-east-1d. The region should be determined by stripping the last character, e.g. us-east-1.
- Check the environment variables AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY and the optional AWS_SESSION_TOKEN for the AWS security credentials. If found, skip using the AWS metadata server to determine these values.